lib/src/where_type.dart

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+/// Emits only the events which have type [R].
+///
+/// If the source stream is a broadcast stream the result will be as well.
+///
+/// Errors from the source stream are forwarded directly to the result stream.
+StreamTransformer<Null, R> whereType<R>() => _WhereType<R>();
+
+class _WhereType<R> extends StreamTransformerBase<Null, R> {
+  @override
+  Stream<R> bind(Stream<Object> values) {
+    var controller = values.isBroadcast
+        ? StreamController<R>.broadcast(sync: true)
+        : StreamController<R>(sync: true);
+
+    StreamSubscription<Object> subscription;
+    controller.onListen = () {
+      if (subscription != null) return;
+      var valuesDone = false;
+      subscription = values.listen(
+          (value) {
+            if (value is R) controller.add(value);
+          },
+          onError: controller.addError,
+          onDone: () {
+            valuesDone = true;
+            controller.close();
+          });
+      if (!values.isBroadcast) {
+        controller.onPause = subscription.pause;
+        controller.onResume = subscription.resume;
+      }
+      controller.onCancel = () {
+        var toCancel = subscription;
+        subscription = null;
+        if (!valuesDone) return toCancel.cancel();
```

Comment:
It seems like it would be cleaner to null out the `subscription` field when the values are done, and get rid of this `valuesDone` field entirely.

Reply:
I think I was worried about the `onListen` getting called again in a broadcast stream after the values where already done and the `subscription != null` condition not blocking us from doing work we shouldn't - but it looks like maybe `StreamController` handles this for us and won't call `onListen` again anyway. Done